google/datalab/contrib/mltoolbox/commands/_ml.py

```diff
@@ -410,7 +414,19 @@ def _transform(args, cell):
   if '--project-id' not in cmd_args:
     cmd_args.extend(['--project-id', google.datalab.Context.default().project_id])
 
-  _shell_process.run_and_monitor(cmd_args, os.getpid(), cwd=MLTOOLBOX_CODE_PATH)
+  try:
+    tmpdir = None
```

Comment:
I think we also need to hash the file, as the user could download the file and use a local path (run 1 notebook with 'file.tar.gz' from gs://bucket1/file.tar.gz, and another notebook at the same time  with 'file.tar.gz' form gs://bucket2/file.tar.gz).

This also doesn't address running the same package from two notebooks at the same time. When the 2nd notebook starts, it shouldn't overwrite files in the first one. Making a cache system is also a partial solution here. For example, what if the analyze.py script writes a file in dir_path(__file__).  The first notebook puts it in the cache, the 2nd notebook sees it already exists, and just runs it. So now two processes are writing to the same file.



Maybe we can make a cache.

All